### DISCUSSION TITLE
intro

### DISCUSSION THREAD

**Discussion by giscus** [isno/theByteBook#150] [General]
# intro

构建大规模高可用的分布式系统

https://www.thebyte.com.cn/intro.html

**Comment by yankaiqian** [comment]
「视角到转回到工程师个体」，这句话中可能有错别字，或者您可以考虑去掉一个「到」字，改为「视角转回到工程师个体」。

**Reply by isno** [reply]
谢谢提醒，已修正。

**Comment by SongJian-99** [comment]
占座学习

**Reply by LinL0U** [reply]
占座学习

**Comment by yujijiwawa** [comment]
“教条式介绍睡过一觉不会有多少人记起。”这句话建议加个标点符号分隔下，或者换个说法

**Reply by isno** [reply]
谢谢，【如何阅读】这一段没组织好，比较杂也很乱，会重新写。

**Comment by FlipThatCover** [comment]
在今天的简中互联网上还能看到这么良心的内容，泪目了。
另外，小弟2024年毕业，上一段实习从事的是链路追踪，容器观测等工作，用到的技术栈有OpenTelemetry、Prometheus、Jaeger、eBPF等。
小弟我今天刚被公司狠狠卡转正，但是对容器可观测技术以及容器化、容器编排很有兴趣，今天了解了下，这类工作可以归结为SRE或者DevOps
想问问已经在行内的大哥们，这条道适合应届生吗，想做的话从哪里做起呢？

**Reply by isno** [reply]
适合啊，你都找到这本书了，就从这本书开始呗。

**Comment by hexody** [comment]
每个章节都太独立了，看完了整本仍然没有完整的“高可用架构原理与实践”知识体系，虽然每个章节都或多或少作为高可用架构的一部分，但是很难有系统性的理解。不过内容还是不错的

**Reply by hexody** [reply]
有两个建议，
1. 如果还按照原书名，最好加个章节将之前章节的内容串起来形成体系，再配一个系统性的案例
2. 索性就改个书名，避开“架构”“实践”这两个词。。架构意味着体系，且书中实践涉及的太少，很难说能有指导意义。比如叫“深入高可用系统原理与设计”

**Reply by isno** [reply]
谢谢建议。我在最近改稿中也注意到了这个问题。

我原本的写作思路是这样的：

-  第一章是系统设计概论，算是全书的绪论。
-  第二章的主题是“网络请求”，是“高可用系统”设计的起点。
-  第三章，请求进入系统，这一章相当于后续章节的前置，介绍一些高层架构设计中，需要注意的底层技术/原理。
-  第四章，开始进入分布式系统，主题是可扩展性，也就是介绍各类负载均衡技术。
-  第五章，第六章：前面章节介绍的都是无状态应用的设计。现在开始介绍“有状态应用”的设计，这里面有两个核心的主题：事务、以及共识。事务以及共识影响系统的可用性、容错设计。【但目前，感觉这两章没写好】。
- 第七章，第八章，是介绍容器系统相关的设计，介绍分布式系统中，容器是如何通信、调度、被管理的。但 Kubernetes 的问题是仅仅管理容器，容器内业务的通信治理也是个重要的主题。作为新一代的基础设施，容器以及服务网格的作用：是将系统中非业务逻辑全部解耦，下沉到容器、Sidecar 中。

-  第九章，解决了“网络请求”、高可用设计（负载均衡），数据一致性（事务、共识），应用负载（容器和服务网格）。那么整个高可用架构“设计阶段”已经解决。现在开始进入系统“运行阶段”。这个阶段要面临：大规模数据的挑战（成本和海量数据分析），系统的稳定性（监控），（各类观测数据的统一处理， 标准和规范）
-  最后一章，是 devops 的内容，补充应用如何交付、部署。

上面是整部书籍的写作思路，但各个章节我没串联起来。我现在的想法是：补充章结论，在章结论中承前启后，引入下一章的主题。

目前，我的感觉是整部书的骨架搭建好了，但细节还是还有太多要补充。我从 2022年，5月左右开始写，到现在，2年半的时间。说长不长，说短也不短，但迟迟完不了稿（总感觉有缺陷，要补充）。给我的压力也很大。

最后，谢谢你的建议，是我现在遇到的”内容困境“，我再想想怎么写。

**Reply by isno** [reply]
”深入高可用系统原理与设计“，这个名字可以。

**Reply by hexody** [reply]
加油加油，追更一年了快

**Reply by hexody** [reply]
这些章节或既然或多少作为”高可用系统“理论基础，且作用在不同的架构层级当中，我个人觉得应该开局先拿一个典型的高可用系统的架构图来先给读者建立一个总体的概念，然后分层自下而上或者自顶向下安排章节，这样比较顺理成章。另外就像你说的，在章节末加一些承前启后的文字过渡一下，也是个比较好的方式。

**Reply by hexody** [reply]
然后全书其实也不完全是在聚焦“高可用”来讲的，如果读者是奔着“高可用”三个字去买这本书我觉得看完之后大概率不会符合他的预期。
本书更像是一个DevOps手册（因为现代的DevOps本身就离不开高可用系统的这些concept），内容偏general，所以在不大改内容的前提下，书名还得斟酌一下🤣

**Reply by isno** [reply]
> 然后全书其实也不完全是在聚焦“高可用”来讲的，如果读者是奔着“高可用”三个字去买这本书我觉得看完之后大概率不会符合他的预期。 本书更像是一个DevOps手册（因为现代的DevOps本身就离不开高可用系统的这些concept），内容偏general，所以在不大改内容的前提下，书名还得斟酌一下🤣

对于“高可用”，我有个更宏观的解释，不仅仅是”不宕机“，还有：
- 成本可用（资源成本，K8S 弹性，海量规模下的日志存储、分析，技术复杂度成本：ServiceMesh 降低业务复杂度）
- 敏捷可用（效率成本，gitops）
- 规模变大，但技术可控（不可变基础设施、容器、容器编排，等等一系列的云原生技术）

你提到的，”典型的高可用系统的架构图“，是否是”两地三中心“、”主备模式“、”集群副本模式“这样的内容？我刻意没写这些内容。我在书序里面表达了一个意思：”技术/架构/方法论“一直在变化，写一写具体的”技术方案“，很容易过时。

但一些基础的理论/原理，不会”过时“，这些理论/原理，是上层各类高可用架构的基石。上层的技术再怎么演进，基石就一直在那里。

我写作手法还不行，上述内容没很好表达出来。

**Comment by CowSmiles** [comment]
"尝试悟透点技术发展的规律。"这个里面的点是不是应该去掉，应该是多余的字

**Reply by isno** [reply]
“悟透技术发展的规律”，这样描述过于自大。加个“点”，只悟透“一点”，得谦虚一下。

**Comment by cypggs** [comment]
希望加个下一页，上一页，不然太难受了。

**Comment by iwangzhy** [comment]
> 《软件在吞噬世界》节选
> 我们处于戏剧性和广泛的技术和经济转变的中间，软件公司准备接管大量的经济。... 十年前，当我在创办的 Netscape 公司时，大概只有 500 万人使用宽带互联网，而现在有超过 20 亿人使用宽带互联网...

https://a16z.com/why-software-is-eating-the-world/  原文说的是  
> Over two billion people now use the broadband Internet, up from perhaps 50 million a decade ago, when I was at Netscape, the company I co-founded. 

应该是 5000万人。
